src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/ConsiderPassingBaseTypesAsParameters.cs

```diff
@@ -0,0 +1,227 @@
+                var eventArgsTypeSymbol = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemEventArgs);
+
+                context.RegisterOperationBlockStartAction(context =>
+                {
+                    if (context.OwningSymbol is not IMethodSymbol methodSymbol ||
+                        methodSymbol.IsOverride ||
+                        methodSymbol.Parameters.Length == 0 ||
+                        methodSymbol.IsImplementationOfAnyInterfaceMember() ||
+                        methodSymbol.HasEventHandlerSignature(eventArgsTypeSymbol) ||
+                        !context.Options.MatchesConfiguredVisibility(Rule, methodSymbol, context.Compilation,
+                            defaultRequiredVisibility: SymbolVisibilityGroup.All) ||
+                        context.Options.IsConfiguredToSkipAnalysis(Rule, methodSymbol,
+                            context.Compilation))
+                    {
+                        return;
+                    }
+
+                    var typesUsedPerParameter = PooledConcurrentDictionary<IParameterSymbol, PooledConcurrentSet<ITypeSymbol>>.GetInstance(SymbolEqualityComparer.Default);
+                    foreach (var param in methodSymbol.Parameters)
+                    {
+                        typesUsedPerParameter.AddOrUpdate(param, PooledConcurrentSet<ITypeSymbol>.GetInstance(SymbolEqualityComparer.Default), (param, value) => value);
+                    }
+
+                    context.RegisterOperationAction(context =>
+                    {
+                        var paramReference = (IParameterReferenceOperation)context.Operation;
+
+                        switch (paramReference.Parent.Kind)
+                        {
+                            case OperationKind.Invocation:
+                                AddDeclaringTypeSymbol(((IInvocationOperation)paramReference.Parent).TargetMethod);
+                                break;
+
+                            case OperationKind.PropertyReference:
+                                AddDeclaringTypeSymbol(((IPropertyReferenceOperation)paramReference.Parent).Property);
+                                break;
+
+                            case OperationKind.EventReference:
+                                AddDeclaringTypeSymbol(((IEventReferenceOperation)paramReference.Parent).Event);
+                                break;
+
+                            case OperationKind.FieldReference:
+                                AddType(((IFieldReferenceOperation)paramReference.Parent).Field.ContainingType);
+                                break;
+
+                            case OperationKind.Argument:
+                                AddType(paramReference.Type);
+                                break;
+
+                            case OperationKind.Conversion:
+                                AddType(paramReference.Parent.Type);
+                                break;
+
+                            default:
```

Comment:
Not sure if you want to keep the fallback case and be explicit about the fact we do nothing?